contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,816 @@
+        volumeMounts:
+        - name: www
+          mountPath: /usr/share/nginx/html
+  volumeClaimTemplates:
+  - metadata:
+      name: www
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+Phased roll outs can be used to roll out a configuration, image, or resource 
+update to some portion of the fleet maintained by the StatefulSet prior to 
+updating the entire fleet. It is useful to support linear, geometric, and 
+exponential roll out of an update. Users can modify the 
+`.Spec.UpdateStrategy.Partition.Ordinal`  to allow the roll out to progress.
+
+```yaml
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: web
+spec:
+  serviceName: "nginx"
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      updateStrategy: 
+        type: Partitioned
+        partition: 
+          ordinal: 1
+      containers:
+      - name: nginx
+        image: gcr.io/google_containers/nginx-slim:0.9
+        ports:
+        - containerPort: 80
+          name: web
+        volumeMounts:
+        - name: www
+          mountPath: /usr/share/nginx/html
+  volumeClaimTemplates:
+  - metadata:
+      name: www
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+### Rollbacks
+To rollback an update, users can use the `kubectl rollout` command.
+
+The command below will roll back the `web` StatefulSet to the previous revision in 
+its history. If a roll out is in progress, it will stop deploying the target 
+revision, and roll back to the current revision.
+
+```shell
+kubectl rollout undo statefulset web
+```
+
+### Rolling Forward
+Rolling back is usually the safest, and often the fastest, strategy to mitigate
+deployment failure, but rolling forward is sometimes the only practical solution 
+for stateful applications (e.g. A users has a minor configuration error but has 
+already modified the storage format for the application). Users can use 
+sequential `kubectl apply`'s to update the StatefulSet's current 
+[target state](#target-state). The StatefulSet's `.Spec.GenerationPartition` 
+will be respected, and it therefore interacts well with canaries and phased roll
+ outs.
+
+## Tests
+- Updating a StatefulSet's containers will trigger updates to the StatefulSet's 
+Pods respecting the 
+[identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity) 
+and [deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+guarantees.
+- A StatefulSet update will block on failure.
+- A StatefulSet update can be rolled back.
+- A StatefulSet update can be rolled forward by applying another update. 
```

Comment:
What is the behavior of the update when during an update (currentRevision=1, updatedRevision=2) I update the PodTemplateSpec to result in 3?